pfb/test/test_wavelets.py

```diff
@@ -0,0 +1,275 @@
+@pytest.mark.parametrize("data_shape", [(13,), (12, 7)])
+def test_dwt_idwt_axis(wavelet, data_shape):
+    pywt = pytest.importorskip("pywt")
+    data = np.random.random(size=data_shape)
+    pywt_dwt_axis = pywt._dwt.dwt_axis
+    pywt_idwt_axis = pywt._dwt.idwt_axis
+
+    pywt_wavelet = pywt.Wavelet(wavelet)
+    pywt_mode = pywt.Modes.symmetric
+
+    wavelet = discrete_wavelet(wavelet)
+
+    for axis in reversed(range(len(data_shape))):
+        # Deconstruct
+        ca, cd = dwt_axis(data, wavelet, Modes.symmetric, axis)
+        pywt_ca, pywt_cd = pywt_dwt_axis(data, pywt_wavelet, pywt_mode, axis)
+        assert_array_almost_equal(ca, pywt_ca)
+        assert_array_almost_equal(cd, pywt_cd)
+
+        # Reconstruct with both approximation and detail
+        pywt_out = pywt_idwt_axis(ca, cd, pywt_wavelet, pywt_mode, axis)
+        output = idwt_axis(ca, cd, wavelet, Modes.symmetric, axis)
+        assert_array_almost_equal(output, pywt_out)
+
+        # Reconstruct with approximation only
+        pywt_out = pywt_idwt_axis(ca, None, pywt_wavelet, pywt_mode, axis)
+        output = idwt_axis(ca, None, wavelet, Modes.symmetric, axis)
+        assert_array_almost_equal(output, pywt_out)
+
+        # Reconstruct with detail only
+        pywt_out = pywt_idwt_axis(None, cd, pywt_wavelet, pywt_mode, axis)
+        output = idwt_axis(None, cd, wavelet, Modes.symmetric, axis)
+        assert_array_almost_equal(output, pywt_out)
+
+
+def test_dwt_idwt():
+    pywt = pytest.importorskip("pywt")
+    data = np.random.random((5, 8, 11))
+
+    res = dwt(data, "db1", "symmetric")
+    pywt_res = pywt.dwtn(data, "db1", "symmetric")
+    for k, v in res.items():
+        assert_array_almost_equal(v, pywt_res[k])
+
+    res = dwt(data, "db1", "symmetric", 1)
+    pywt_res = pywt.dwtn(data, "db1", "symmetric", (1,))
+    for k, v in res.items():
+        assert_array_almost_equal(v, pywt_res[k])
+
+    res = dwt(data, ("db1", "db2"), ("symmetric", "symmetric"), (0, 1))
+    pywt_res = pywt.dwtn(data, ("db1", "db2"), ("symmetric", "symmetric"), (0, 1))
+    for k, v in res.items():
+        assert_array_almost_equal(v, pywt_res[k])
+
+    output = idwt(res, ("db1", "db2"), ("symmetric", "symmetric"), (0, 1))
+    pywt_out = pywt.idwtn(pywt_res, ("db1", "db2"), ("symmetric", "symmetric"), (0, 1))
+    assert_array_almost_equal(output, pywt_out)
+
+
+@pytest.mark.parametrize("data_shape", [(50, 24, 63)])
+def test_wavedecn_waverecn(data_shape):
+    pywt = pytest.importorskip("pywt")
+    data = np.random.random(data_shape)
+
+    out = pywt.wavedecn(data, "db1", "symmetric")
+    a, coeffs = wavedecn(data, "db1", "symmetric")
+
+    assert_array_almost_equal(a, out[0])
+
+    for d1, d2 in zip(out[1:], coeffs):
+        assert list(d1.keys()) == list(d2.keys())
+
+        for k, v in d1.items():
+            assert_array_almost_equal(v, d2[k])
+
+    pywt_rec = pywt.waverecn(out, "db1", "symmetric")
```

Comment:
I usually use this with mode='periodization' but I see below that it has not been implemented. I don't know how big a difference this makes. Will have to test

Reply:
Would it not be possible to use zeropad, in light of email conversations?